Introduce new decora_ble integration

ColinCampbell opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This pull request introduces a modern Decora BLE integration. There is an existing `decora` integration which is years out-of-date (uses bluepy, for example). I would happily replace the existing integration if that's the right thing to do here. One complication is that the `decora` integration required manual configuration in `configuration.yaml` and I wasn't sure how to port from that to the config_flow system.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

I'm going to implement reauth in the config flow to allow for updating the api key in a followup.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28245

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ColinCampbell committed:
Introduce new decora_ble integration

bdraco commented on homeassistant/components/decora_ble/__init__.py:
```diff
@@ -0,0 +1,82 @@
+    api_key: str = entry.data[CONF_API_KEY]
+
+    ble_device = bluetooth.async_ble_device_from_address(
+        hass, address.upper(), connectable=True
```
Store the address as `upper()` so you don't have to do this every time

bdraco commented on homeassistant/components/decora_ble/manifest.json:
```diff
@@ -0,0 +1,19 @@
+  "integration_type": "device",
+  "iot_class": "local_push",
+  "requirements": ["decora_bleak==0.1.0"],
+  "ssdp": [],
+  "zeroconf": []
```
```suggestion
```

bdraco commented on homeassistant/components/decora_ble/manifest.json:
```diff
@@ -0,0 +1,19 @@
+  "config_flow": true,
+  "dependencies": ["bluetooth_adapters"],
+  "documentation": "https://www.home-assistant.io/integrations/decora_ble",
+  "homekit": {},
```
```suggestion
```

bdraco commented on homeassistant/components/decora_ble/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+
+    async def _attempt_create_entry(
+        self, address: str, name: str
+    ) -> Union[FlowResult, str]:
```
Please avoid progressing the flow in the helper.

Make this return an error string or None and let the caller progress the flow based on that.

bdraco commented on homeassistant/components/decora_ble/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+                    title=name, data={CONF_ADDRESS: address, CONF_API_KEY: api_key}
+                )
+
+            _LOGGER.error("Device did not return an API key, not in pairing mode")
+            return "not_in_pairing_mode"
+
+        _LOGGER.error("Could not find BLE device")
```
```suggestion
            return "not_in_pairing_mode"

```

These likely won't be seen by the user and generate log noise. Its better for it to be displayed in the UI

bdraco commented on homeassistant/components/decora_ble/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
+
+DEFAULT_NAME = "Room Lights"
```
It would be better to store nothing instead of a default here so we know in the future if its unset in the event we can somehow get the name for the device and use it when the user didn't set it before.

ColinCampbell committed:
Remove test that was copied over but improperly updated, unnecessary

emontnemery commented:
@ColinCampbell please address the review comments, then click the "Ready for review" button 👍

ColinCampbell committed:
Remove superfluous items from manifest.json

ColinCampbell committed:
Clarify API key fetching so progress is determined in the main flow methods

ColinCampbell committed:
Store address as upper so we don't need to do it when reading config value

ColinCampbell committed:
Remove default name for entity

ColinCampbell commented:
Thank you for the review bdraco! Addressed the issues you called out.

emontnemery commented:
@ColinCampbell Please open a documentation PR and link to it from this PR. Make sure the documentation PR is based on and targets the `next` branch.

ColinCampbell commented:
Documentation and brands PRs have been sent and added here

IsakNyberg commented on homeassistant/components/decora_ble/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+
+        if user_input is not None:
+            api_key, error = await self._get_api_key(address)
+            if api_key is not None:
+                assert error is None
```
```suggestion
            if api_key is not None and error is None:
```

IsakNyberg commented on homeassistant/components/decora_ble/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+            self._abort_if_unique_id_configured()
+
+            api_key, error = await self._get_api_key(address)
+            if api_key is not None:
+                assert error is None
```
```suggestion
            if api_key is not None and error is None:
```

IsakNyberg reviewed (commented):
Just some small quality suggestions

ColinCampbell committed:
Adjustments around checking values and early returns

bdraco commented on homeassistant/components/decora_ble/config_flow.py:
```diff
@@ -0,0 +1,150 @@
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
```
Please remove this.  Debug logging should be controlled by the logger integration

bdraco commented on homeassistant/components/decora_ble/light.py:
```diff
@@ -0,0 +1,100 @@
+    @callback
+    def _async_decora_connected(self, summary: DecoraBLEDeviceSummary) -> None:
+        """Handle state changed."""
+        self._attr_device_info = DeviceInfo(
```
Set this in __init__.  It's too late to set it here as it's read when the entity is added to hass

bdraco commented on homeassistant/components/decora_ble/light.py:
```diff
@@ -0,0 +1,100 @@
+            connections={(dr.CONNECTION_BLUETOOTH, self._address)},
+        )
+
+        self.schedule_update_ha_state()
```
```suggestion
        self.async_write_ha_state()
```

ColinCampbell committed:
Get rid of superfluous logger calls at the top of decora_ble config_flow

ColinCampbell committed:
Updates to properly wait for summary state to be available prior to creating the entry

MartinHjelmare commented:
Don't we want to replace the existing integration and reuse that domain and import the configuration.yaml to the config flow via an import step?

bdraco commented:
> Don't we want to replace the existing integration and reuse that domain and import the configuration.yaml to the config flow via an import step?

I think that would be a good idea since I don't think the current one works at all given https://www.home-assistant.io/integrations/decora has 0 installs

ColinCampbell commented:
Are there examples of integrations that do that? I thought about doing trying to reuse the existing integration. My concern is that at some point between the old integration being written and now, Leviton changed the format of the API key the device firmware expects, and the API key is stored in the old `configuration.yaml` files. Anything ported from the old integration to this new one would need to be re-added or at the very least re-authenticated. That, and the fact that there were 0 installs of the old integration, I wasn't sure that the additional effort to try and port over existing usages would be worth it.

bdraco commented:
How does the user get the api key with the old method vs the new method?

Is there a way the old key can be converted into the new format?

ColinCampbell commented:
The switch is put into a pairing mode by pressing and hold down on the switch paddle for a few seconds, then the system communicating with it over Bluetooth is able to ask for the API key. As far as I know, there's no conversion method to go from old API keys to new ones. My devices were already on the most recent version of the firmware when I bought them and everything I was able to find about connecting to the devices since the firmware upgrade required repairing the devices.

MartinHjelmare commented:
Maybe import the device data to config entries without validating the api key, and then when the config entry is setup it will fail authentication and a reauth flow can be started which will ask the user to put the device in pairing mode to re-authenticate?

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

You can search for `async_step_import` in the repo for integrations that import data via the config flow to config entries.

ColinCampbell committed:
Introduce DiscoveredDecoraDevice to abstract the bluetooth discovery info to prepare to allow for importing from configuration.yaml

ColinCampbell committed:
More descriptive names for the results in the config flow tests

ColinCampbell committed:
Merge into diff creating DiscoveredDecoraDevice

ColinCampbell commented:
I've started implementing the import and reauth steps! I'm travelling this week so it might be a bit before an update.

Another question, I'm not sure I'll be able to get my device downgraded to past firmware to check that the new bleak-based package I wrote works with previous firmware versions. I can double check the code in communicating with the switches, but is there any prior art and/or hard or soft rules on changing integrations that require upgrading device firmware?

emontnemery commented:
> I've started implementing the import and reauth steps! I'm travelling this week so it might be a bit before an update.

I've set the PR as draft pending your update. Please don't forget to click the "ready for review" button once you've implemented import of old configurations.

> I can double check the code in communicating with the switches, but is there any prior art and/or hard or soft rules on changing integrations that require upgrading device firmware?

If the firmware us user upgradeable, I don't see any problems with requiring the user to do a firmware upgrade. If you can't test it yourself, I'd suggest to block old firmware already in the config flow if possible, and just ask the user to upgrade the device.

ColinCampbell committed:
Add support for importing old decora integration configs

ColinCampbell committed:
Upgrade decora_bleak to v0.2.2

ColinCampbell committed:
Replace the Decora integration with the newly created Decora BLE code

ColinCampbell commented:
Ok so I think this is ready for review again :) Sorry for the delays. I've implemented the changes to (a) move the new code into the main decora integration and (b) implement the import of the configuration.yaml based config into the ConfigFlow based config. I have config_flow.py fully unit tested but I had a hard time testing the import flow manually. I added the config to the configuration.yaml but I couldn't get it to trigger the import flow as far as I could test but it did toss an error out saying that the configuration needed to move to the Config Flow based version. I'm travelling this weekend but I'll take a peek at testing this further when I'm back on Monday.

emontnemery commented:
Tests are failing, please fix the tests and mark the PR ready for review when done.

Maybe the tests are not patching the bluetooth functionality so the tests try to do actual bluetooth access?
This is just a guess from my side though.
